src/app/components/queue/form/QuestionForm.tsx

```diff
@@ -0,0 +1,209 @@
+import {
+  Box,
+  Button,
+  Checkbox,
+  Drawer,
+  FormControl,
+  FormControlLabel,
+  FormLabel,
+  TextField,
+} from "@mui/material";
+import { IoMdClose } from "react-icons/io";
+import React from "react";
+import { officeHourContext } from "@context/OfficeHourContext";
+import { MultipleChoiceTags, SingleChoiceTags } from "./Tags";
+import { TagOption } from "@interfaces/db";
+import { createQuestion } from "@utils/index";
+import { serverTimestamp } from "firebase/firestore";
+import { UserSessionContext } from "@context/UserSessionContext";
+import { IdentifiableQuestion } from "@interfaces/type";
+
+interface QuestionFormProps {
+  // button to open the form
+  triggerButton: JSX.Element;
+  title: string;
+  // current state of question
+  currentQuestion: IdentifiableQuestion;
+}
+
+const QuestionForm = (props: QuestionFormProps) => {
+  const { triggerButton, title, currentQuestion } = props;
+  const [openForm, setOpenForm] = React.useState(false);
+  const ohContext = React.useContext(officeHourContext);
+  const userSessionContext = React.useContext(UserSessionContext);
+
+  const [question, setQuestion] = React.useState(currentQuestion.title);
+  const [description, setDescription] = React.useState(
+    currentQuestion.description
+  );
+  const [questionPublic, setQuestionPublic] = React.useState(
+    currentQuestion.questionPublic
+  );
+
+  const [questionTags, setQuestionTags] = React.useState<
+    Record<string, TagOption[]>
+  >(() => {
+    const curr_tags = currentQuestion.tags.map((t) => t.choice);
+    let init: Record<string, TagOption[]> = {};
+    Object.keys(ohContext.course.tags).forEach((t) => {
+      init[t] = ohContext.course.tags[t].options.filter((o) =>
+        curr_tags.includes(o.choice)
+      );
+    });
+    return init;
+  });
+
+  const updateQuestionTags = (tagsKey: string, newTags: TagOption[]) => {
+    setQuestionTags({ ...questionTags, [tagsKey]: newTags });
+  };
+
+  const trigger = React.cloneElement(triggerButton, {
+    onClick: () => {
+      setOpenForm(true);
+    },
+  });
+
+  const joinQueue = () => {
+    const timestamp = serverTimestamp();
+    const author = userSessionContext.user?.tufts_username;
+
+    let tagsArr = Object.values(questionTags).reduce((arr, curr) => {
+      return [...arr, ...curr];
+    }, []);
+
+    const tagsValidate = Object.keys(questionTags).filter(
+      (k) => ohContext.course.tags[k].required && questionTags[k].length === 0
+    );
+
+    if (question !== "" && description !== "" && tagsValidate.length === 0) {
+      if (title === "Join queue") {
+        createQuestion(
```

Comment:
We will probably use `react-hook-form` or a similar library for field validation, so we can omit this check